### PR TITLE
Add Presto iceberg connector

### PR DESCRIPTION
**Pull request by ChunxuTang**
Cherry-pick of https://github.com/trinodb/trino/commit/e82c2d5301396ca16248eb6931bec11bf1352470

Co-Authored-By: Parth Brahmbhatt <pbrahmbhatt@netflix.com>
Co-Authored-By: Zhenxiao Luo <zluo@twitter.com>
Co-Authored-By: Beinan Wang <beinanw@twitter.com>

```
== RELEASE NOTES ==

General Changes
* Add iceberg connector
```

### CODE REVIEW

**Comment by beinan** [comment]
Hello @ChunxuTang could you rebase master? Thanks!

**Comment by ChunxuTang** [comment]
#15836

**Comment by ChunxuTang** [comment]
#15620

**Comment by ChunxuTang** [comment]
Provided a quick fix for the parquet conflicts to kick-off the review and testing. Also sent a PR to the prestodb/presto-hive-apache repo for a parquet upgrade. When that PR is merged, we can remove the quick fix safely.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSessionProperties.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                        false,
+                        value -> HiveCompressionCodec.valueOf(((String) value).toUpperCase()),
+                        HiveCompressionCodec::name),
+                booleanProperty(
+                        PARQUET_FAIL_WITH_CORRUPTED_STATISTICS,
+                        "Parquet: Fail when scanning Parquet files with corrupted statistics",
+                        hiveClientConfig.isFailOnCorruptedParquetStatistics(),
+                        false),
+                booleanProperty(
+                        PARQUET_USE_COLUMN_NAMES,
```
Comment (beinan) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSessionProperties.java:66]: Do we really use this in iceberg?

File: presto-spark-base/pom.xml [hunk 2]
```diff
@@ -245,6 +245,10 @@
                         <ignoredResourcePattern>parquet.thrift</ignoredResourcePattern>
                         <ignoredResourcePattern>about.html</ignoredResourcePattern>
                     </ignoredResourcePatterns>
+                    <ignoredClassPatterns>
```
Comment (beinan) [inline presto-spark-base/pom.xml:248]: is this change still needed after we update presto-hive-apache?  if not, add a todo or something?

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergDistributed.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        return false;
+    }
+
+    public void testJoinWithLessThanOnDatesInJoinClause()
+    {
+    }
+
+    @Override
+    public void testRenameTable()
+    {
```
Comment (beinan) [inline presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergDistributed.java:52]: leave a comment if the rename table feature has not been supported yet,  or add a todo if it's already supported

**Review by beinan — COMMENTED** [review]
Great work! @ChunxuTang Just put a few questions and code style issues

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 4]
```diff
@@ -1330,7 +1330,7 @@ private static Table buildTableObject(
         return tableBuilder.build();
     }
 
-    private static PrincipalPrivileges buildInitialPrivilegeSet(String tableOwner)
+    public static PrincipalPrivileges buildInitialPrivilegeSet(String tableOwner)
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Good catch! Have changed it back to `private`.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HdfsFileIo.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+
+import java.io.IOException;
+
+import static com.facebook.presto.iceberg.IcebergErrorCode.ICEBERG_FILESYSTEM_ERROR;
+import static java.util.Objects.requireNonNull;
+
+public class HdfsFileIo
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HdfsFileIo.java]: Yes, using the upper case of 'O' is better here. Refactored the code here.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+                metastore,
+                database,
+                table,
+                Optional.of(requireNonNull(owner, "owner is null")),
+                Optional.of(requireNonNull(location, "location is null")));
+    }
+
+    private HiveTableOperations(FileIO fileIo, ExtendedHiveMetastore metastore, String database, String table, Optional<String> owner, Optional<String> location)
+    {
+        this.fileIo = requireNonNull(fileIo, "fileIo is null");
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java]: Also updated~

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+        String newMetadataLocation = writeNewMetadata(metadata, version + 1);
+
+        // TODO: use metastore locking
+
+        Table table;
+        try {
+            if (base == null) {
+                String tableComment = metadata.properties().get(TABLE_COMMENT);
+                Map<String, String> parameters = new HashMap<>();
+                parameters.put("EXTERNAL", "TRUE");
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java:185]: Thanks for the suggestion! I've gone through properties defined in iceberg, e.g. `BaseMetastoreTableOperations`. But I'm afraid that I didn't find a constant to reuse here.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+    private static int parseVersion(String metadataLocation)
+    {
+        int versionStart = metadataLocation.lastIndexOf('/') + 1; // if '/' isn't found, this will be 0
+        int versionEnd = metadataLocation.indexOf('-', versionStart);
+        try {
+            return parseInt(metadataLocation.substring(versionStart, versionEnd));
+        }
+        catch (NumberFormatException | IndexOutOfBoundsException e) {
+            log.warn(e, "Unable to parse version from metadata location: %s", metadataLocation);
+            return -1;
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java:352]: Thanks for bringing it up! We don't fully support versions for now. When there's an error in parsing the version, it means the version has been polluted and will be meaningless. That's why we just return -1 here.

I'm thinking of adding more control or returning more information around the version in the future. For now, I think we can leave it as it is now. Do you have any suggestions?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    public ExecutorService createCachingHiveMetastoreExecutor(MetastoreClientConfig metastoreClientConfig)
+    {
+        return newFixedThreadPool(
+                metastoreClientConfig.getMaxMetastoreRefreshThreads(),
+                daemonThreadsNamed("hive-metastore-iceberg-%s"));
+    }
+
+    @Singleton
+    @Provides
+    public CacheManager createCacheManager(CacheConfig cacheConfig, FileMergeCacheConfig fileMergeCacheConfig, CacheStats cacheStats)
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java:133]: Have checked the `CachingModule.java` file. The `createCacheManager` functions are different in these two classes with different threads (created via thread pool). So I think we cannot reuse that module here.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSessionProperties.java (hunk 1 above)
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSessionProperties.java:66]: Yes, it is used in the `createDataPageSource` function in the `IcebergPageSourceProvider.java` file~

File: presto-spark-base/pom.xml (hunk 2 above)
Comment (ChunxuTang) [reply presto-spark-base/pom.xml:248]: Nice catch! Yeah, I think we may still need this after updating the parquet in the presto-hive-apache to avoid dependency conflicts. We can do more tests of compilation to confirm that, avoiding potential changes in other packages' POM files.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergDistributed.java (hunk 3 above)
Comment (ChunxuTang) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergDistributed.java:52]: Yeah, we don't support it yet. The reason we have the function is that we need to override the function defined in the `AbstractTestDistributedQueries`. We leverage that class to test SQL queries against the connector.

**Review by zhenxiao — CHANGES_REQUESTED** [review]
nice work, @ChunxuTang 
added a few comments. could you please add cherry-pick info, and original commit in commit message?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java:352]: I'm good to use -1 here now

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java:185]: cool, no worry about it.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (beinan) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java:133]: Got it, let's keep the current implementation.

**Review by beinan — APPROVED** [review]

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+            catch (RuntimeException ex) {
+                e.addSuppressed(ex);
+            }
+            throw e;
+        }
+
+        PrestoPrincipal owner = new PrestoPrincipal(USER, table.getOwner());
+        PrincipalPrivileges privileges = new PrincipalPrivileges(
+                ImmutableMultimap.<String, HivePrivilegeInfo>builder()
+                        .put(table.getOwner(), new HivePrivilegeInfo(HivePrivilege.SELECT, true, owner, owner))
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java]: Sure, updated in the new commit.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+                                METADATA_LOCATION, newMetadataLocation,
+                                PREVIOUS_METADATA_LOCATION, currentMetadataLocation))
+                        .build();
+            }
+        }
+        catch (RuntimeException e) {
+            try {
+                io().deleteFile(newMetadataLocation);
+            }
+            catch (RuntimeException ex) {
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java]: Name updated. Thanks for figuring it out!

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergMetadata.java [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
+        for (SchemaTableName table : tables) {
+            try {
+                columns.put(table, getTableMetadata(session, table).getColumns());
+            }
+            catch (TableNotFoundException e) {
+                // table disappeared during listing operation
+            }
+            catch (UnknownTableTypeException e) {
+                // ignore table of unknown type
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergMetadata.java]: Nice catch! Added extra logging here.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergMetadata.java [hunk 13]
```diff
@@ -0,0 +1,559 @@
...
+    {
+        List<SchemaTableName> tables = listTables(session, Optional.of(prefix.getSchemaName()));
+
+        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
+        for (SchemaTableName table : tables) {
+            try {
+                columns.put(table, getTableMetadata(session, table).getColumns());
+            }
+            catch (TableNotFoundException e) {
+                // table disappeared during listing operation
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergMetadata.java]: Same as above. Added extra logging.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSource.java [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+
+    public IcebergPageSource(
+            List<IcebergColumnHandle> columns,
+            Map<Integer, String> partitionKeys,
+            ConnectorPageSource delegate,
+            TimeZoneKey timeZoneKey)
+    {
+        int size = requireNonNull(columns, "columns is null").size();
+        requireNonNull(partitionKeys, "partitionKeys is null");
+        this.delegate = requireNonNull(delegate, "delegate is null");
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSource.java:72]: The `this.delegate` seems to be necessary here. Removed the `this` keywords for `prefilledBlocks` and `delegateIndexes`.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableName.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+            try {
+                type = TableType.valueOf(typeString.toUpperCase(Locale.ROOT));
+            }
+            catch (IllegalArgumentException e) {
+                throw new PrestoException(NOT_SUPPORTED, format("Invalid Iceberg table name (unknown type '%s'): %s", typeString, name));
+            }
+        }
+
+        Optional<Long> version = Optional.empty();
+        if (type == TableType.DATA || type == TableType.PARTITIONS || type == TableType.MANIFESTS || type == TableType.FILES) {
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableName.java]: Sure, updated~

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableName.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    public static IcebergTableName from(String name)
+    {
+        Matcher match = TABLE_PATTERN.matcher(name);
+        if (!match.matches()) {
+            throw new PrestoException(NOT_SUPPORTED, "Invalid Iceberg table name: " + name);
+        }
+
+        String table = match.group("table");
+        String typeString = match.group("type");
+        String ver1 = match.group("ver1");
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergTableName.java]: Sure, updated the variable names.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        }
+        else {
+            metastore.replaceTable(database, tableName, table, privileges);
+        }
+
+        shouldRefresh = true;
+    }
+
+    @Override
+    public FileIO io()
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java:251]: The `io` function here overrides the base function. The name in inherited from the base class defined in the iceberg.

**Comment by ChunxuTang** [comment]
> nice work, @ChunxuTang
> added a few comments. could you please add cherry-pick info, and original commit in commit message?

Sure~ Added the cherry-pick info in both the PR and the commit message.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergErrorCode.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+    ICEBERG_INVALID_PARTITION_VALUE(3, EXTERNAL),
+    ICEBERG_BAD_DATA(4, EXTERNAL),
+    ICEBERG_MISSING_DATA(5, EXTERNAL),
+    ICEBERG_CANNOT_OPEN_SPLIT(6, EXTERNAL),
+    ICEBERG_WRITER_OPEN_ERROR(7, EXTERNAL),
+    ICEBERG_FILESYSTEM_ERROR(8, EXTERNAL),
+    ICEBERG_CURSOR_ERROR(9, EXTERNAL),
+    ICEBERG_WRITE_VALIDATION_FAILED(10, INTERNAL_ERROR),
+    ICEBERG_INVALID_SNAPSHOT_ID(11, USER_ERROR),
+    /**/;
```
Comment (zhenxiao) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergErrorCode.java:39]: remove `/**/`

**Review by zhenxiao — APPROVED** [review]
looks good to me
thank you @ChunxuTang 
a few minor issues

File: presto-iceberg/pom.xml [hunk 19]
```diff
@@ -0,0 +1,425 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.facebook.presto</groupId>
+        <artifactId>presto-root</artifactId>
+        <version>0.253-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/pom.xml]: Sure~ Rebased the master branch and upgraded the presto-iceberg module to 0.254-snapshot.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergErrorCode.java (hunk 18 above)
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergErrorCode.java]: Thanks for the suggestion! Removed the unnecessary comment.

**Comment by zhenxiao** [comment]
hi @tdcmeehan seems all tests passed, except for Facebook integration. could you please help check if this PR is good to proceed?

**Comment by tdcmeehan** [comment]
The Presto Facebook test failure is unrelated.
